test/node_e2e: wire-in cri-enabled local testing

lucab opened the pull request:
This commit wires-in the pre-existing `--container-runtime` flag for
local node_e2e testing.
This is needed in order to further skip docker specific testing
and validation.

Local CRI node_e2e can now be performed via
`make test-e2e-node RUNTIME=remote REMOTE=false`
which will also take care of passing the appropriate argument to
the kubelet.

Commit:
test/node_e2e: wire-in "container-runtime" for local tests

This commit wires-in the pre-existing `--container-runtime` flag for
local node_e2e testing.
This is needed in order to further skip docker specific testing
and validation.

Local CRI node_e2e can now be performed via
`make test-e2e-node RUNTIME=remote REMOTE=false`
which will also take care of passing the appropriate arguments to
the kubelet.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40525)
<!-- Reviewable:end -->

lucab commented:
/cc @Random-Liu @yujuhong as they did the previous CRI pass on test runners.

lucab commented:
/cc @squeed @euank @yifan-gu for tracking

s-urbaniak commented:
LGTM

yujuhong commented:
@lucab just to make sure I understand what you're trying to do. You want to run local node e2e tests with CRI enabled, but would like to skip the docker-specific testing. Is that correct?

If my understanding is correct, could you base the decision of turning off docker-specific tests by looking at `--container-runtime=docker`? This way, we can still run those tests for the docker-CRI integration.

Random-Liu commented on test/e2e_node/system/validators.go:
```diff
@@ -54,6 +47,21 @@ func Validate(spec SysSpec, report Reporter) error {
+
+	validators := osValidators
+	if !criEnabled {
+		validators = append(validators, dockerValidators...)
```
I think we should run different validation for different runtimes, instead of just skipping it when cri is enabled.

Let's add a flag `container-runtime` instead of the `cri-enabled` flag.

lucab replied:
Ack.

Random-Liu commented on test/e2e_node/e2e_node_suite_test.go:
```diff
@@ -54,6 +54,7 @@ var e2es *services.E2EServices
 var runServicesMode = flag.Bool("run-services-mode", false, "If true, only run services (etcd, apiserver) in current process, and not run test.")
 var runKubeletMode = flag.Bool("run-kubelet-mode", false, "If true, only start kubelet, and not run test.")
 var systemValidateMode = flag.Bool("system-validate-mode", false, "If true, only run system validation in current process, and not run test.")
+var enableCriFlag = flag.Bool("enable-cri", false, "If true, test will run in CRI mode and avoid docker-specific assumptions.")
```
Can we change this to `container-runtime`, and add this to https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/test_context.go#L107?

lucab replied:
Sure I will move into there (sorry I didn't noticed it before). Regarding the name change, `enable-cri` was [already referenced](https://github.com/kubernetes/kubernetes/pull/40525/files#diff-19896e63f8ba932b76e0e277a1c4b4beL68) but I'm, not sure about the history/rationale of it. Should I just go ahead and change all references?

Random-Liu replied:
Sorry, that is a legacy code we missed when refactoring.

Previously, we were also using `enable-cri` to enable and disable cri integration in node e2e. However, later we added the `kubelet-flags` flag, and passed all kubelet flags through it including the `experimental-cri` flag.

lucab commented:
> @lucab just to make sure I understand what you're trying to do. You want to run local node e2e tests with CRI enabled, but would like to skip the docker-specific testing. Is that correct?

Correct.

> If my understanding is correct, could you base the decision of turning off docker-specific tests by looking at --container-runtime=docker? This way, we can still run those tests for the docker-CRI integration.

Sure, I'll do. This goes in the same direction of @Random-Liu suggestion, so I'll wait for a minor clarification from him and then proceed with this. Thanks for the feedback.

lucab commented:
Ok, I've removed last references to `--enable-cri`, moved `--container-runtime` to be available for local testing, introduced a `RUNTIME` make variable and defaulted it to "docker", introduced switch-casing on validators (docker-only right now). PTAL.

tmrts commented:
applying a soft LGTM :+1:

Random-Liu commented:
LGTM

Random-Liu commented:
/approve

tmrts commented:
/approve

deads2k commented:
/approve

runcom commented on hack/make-rules/test-e2e-node.sh:
```diff
@@ -28,6 +28,7 @@ skip=${SKIP-"\[Flaky\]|\[Slow\]|\[Serial\]"}
 parallelism=${PARALLELISM:-8}
 artifacts=${ARTIFACTS:-"/tmp/_artifacts/`date +%y%m%dT%H%M%S`"}
 remote=${REMOTE:-"false"}
+runtime=${RUNTIME:-"docker"}
```
add this to the `Makefile`'s examples https://github.com/kubernetes/kubernetes/blob/master/Makefile#L200 ?

lucab replied:
Ack.

Random-Liu commented:
LGTM, please address @runcom's comment.

lucab commented:
@Random-Liu sure, done.

calebamiles commented:
@k8s-bot ok to test

Random-Liu commented:
/approve
/lgtm

spxtr commented:
@k8s-bot gci gke e2e test this

dchen1107 commented:
/lgtm
and
/approve
